nengo_spa/connectors.py

```diff
@@ -0,0 +1,221 @@
+    """Represents a routed connection from an AST node to a `.ModuleInput`.
+
+    A routed connection is passed through an inhibitable channel to allow the
+    routed connection to be disabled.
+
+    Parameters
+    ----------
+    source : nengo_spa.ast.base.Node
+        The AST node providing the source signal.
+    sink : ModuleInput
+        The module input that the source signal should be routed to.
+    """
+
+    free_floating = set()
+
+    def __init__(self, source, sink):
+        self.type = infer_types(source, sink)
+        self.source = source
+        self.sink = sink
+        RoutedConnection.free_floating.add(self)
+
+    def connect_to(self, sink):
+        return self.source.connect_to(sink)
+
+    def construct(self):
+        return self.connect_to(self.sink.input)
+
+    @property
+    def fixed(self):
+        """Whether the source provides a fixed value."""
+        return isinstance(self.source, Fixed)
+
+    def transform(self):
+        """For a fixed source, returns the transform to implement the output.
+        """
+        assert self.fixed
+        if self.type == TScalar:
+            return self.source.evaluate()
+        else:
+            return np.atleast_2d(self.source.evaluate().v).T
+
+
+class SpaOperatorMixin(object):
+    """Mixin class that implements the SPA operators.
+
+    All operands will be converted to AST node and the implementation of the
+    operator itself is delegated to the implementation provided by those nodes.
+    """
+
+    @staticmethod
+    def __define_unary_op(op):
+        def op_impl(self):
+            return getattr(as_ast_node(self), op)()
+        return op_impl
+
+    @staticmethod
+    def __define_binary_op(op):
+        def op_impl(self, other):
+            return getattr(as_ast_node(self), op)(as_ast_node(other))
+        return op_impl
+
+    __invert__ = __define_unary_op.__func__('__invert__')
+    __neg__ = __define_unary_op.__func__('__neg__')
+
+    __add__ = __define_binary_op.__func__('__add__')
+    __radd__ = __define_binary_op.__func__('__radd__')
+    __sub__ = __define_binary_op.__func__('__sub__')
+    __rsub__ = __define_binary_op.__func__('__rsub__')
+    __mul__ = __define_binary_op.__func__('__mul__')
+    __rmul__ = __define_binary_op.__func__('__rmul__')
+    __matmul__ = __define_binary_op.__func__('__matmul__')
+    __rmatmul__ = __define_binary_op.__func__('__rmatmul__')
+
+    def __rshift__(self, other):
+        return as_ast_node(self) >> as_sink(other)
+
+    def __rrshift__(self, other):
+        return as_ast_node(other) >> as_sink(self)
+
+    dot = __define_binary_op.__func__('dot')
+    rdot = __define_binary_op.__func__('rdot')
+
+    def reinterpret(self, vocab=None):
+        return as_ast_node(self).reinterpret(vocab)
+
+    def translate(self, vocab, populate=None, keys=None, solver=None):
+        return as_ast_node(self).translate(vocab, populate, keys, solver)
+
+
+class ConnectorRegistry(object):
```

Comment:
This code is pretty much copied from the `spa.Network.declare_input` and `spa.Network.declare_output` functions and put into a class to have the code only once (instead of duplicated in both methods). It associates a Nengo object with a vocabulary and allows it to be used in the SPA syntax by dynamically changing the class of the Nengo object instance. The new class is also dynamically generated as a class deriving from the original class of the Nengo object and `SpaOperatorMixin`. Because there is only a limited number of Nengo objects, I don't want to create a new type/class each time and cache them. Furthermore, this ensures that everything with the same type has an identical class.

This is definitely a bit of Python black magic ... so maybe I should add a comment, but not sure yet how to make it concise and helpful.

- [ ] Writing this I noticed that the `_type_cache` should actually be a class variable.